File: dicesimulation.py
#By Jack Tsai

import random


def ran_keep(S2):
    b=S2
    S2=random.randint(1,S2)
    if S2 == 1:
        return 1
    elif S2 == a:
        return 1
    else:
        return 1+ran_keep(S2)

S=int(input(""))
throw=1
avg=0
if S==1:
    avg=throw
else:
    for j in range (10):
        throw=1
        for i in range (10000):
            S1=random.randint(1,S)
            if S1==1:
                throw = throw+1
            else:
                S2=random.randint(1,S1)
                if S2 == 1:
                    throw = throw+2
                else:
                    a=S2
                    S2=random.randint(1,S2)
                    if S2==1:
                        throw = throw+3
                    elif S2==a:
                        throw = throw+3
                    else:
                        throw = throw + 3+ ran_keep(S2)
        average=[round(throw/10000)]
    avg=min(average)
print(avg)

File: dicesimulation.cpp
//By Jack Tsai

#include <iostream>
#include <cmath> 
#include <algorithm>
using namespace std;

int fun(int S2,int a)
{
	S2 = rand() % S2 + 1;
	int b = S2;
	
	if (S2 == a || S2 == 1)
		return 1;
	else
		return 1 + fun(S2, b);

}
int main()
{
	int S, S1, S2;
	cin >> S;
	if(S > 0 && S < 10 ^ 5)
	{
		double avg = 0.0, Throw;
		double average[10];
		if (S == 1)
			avg = 1;
		else
		{
			for (int i = 0; i < 10; i++)
			{
				Throw = 1;
				for (int j = 0; j < 10000; j++)
				{
					S1 = rand() % S + 1;
					if (S1 == 1)
					{
						Throw = Throw + 1;
					}
					else
					{
						S2 = rand() % S1 + 1;
						
						if (S2 == 1)
							Throw = Throw + 2;
						else
						{
							int a = S2;
							S2 = rand() % S2 + 1;
							int b = S2;
							if(S2 == a || S2 == 1)
								Throw = Throw + 3;
							else
								Throw=Throw+3+fun(S2,b);
						}
					}
				}
				average[i] =round( Throw / 10000);
			}
			avg = *min_element(average,average+10);

		}
		cout << avg;
	}
	
	return 0;
}
